proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1160 @@
+
+A function to reorder the list of validators as specified in [LIP 0003][LIP-0003].
+
+
+##### Parameters
+
+The function has the following input parameters in the order given below:
+
+* `validatorsAddresses`: An array of pairwise distinct 20-byte addresses.
+* `randomSeed`: A 32-byte value representing a random seed.
+
+
+##### Returns
+
+This function returns an array of bytes with the re-ordered list of addresses.
+
+
+##### Execution
+
+```python
+shuffleValidatorsList(validatorsAddresses, randomSeed):
+    
+    roundHash = {}
+    for address in validatorsAddresses:
+        roundHash[address] = hash(randomSeed || address)
+
+    # Reorder the validator list
+    shuffledValidatorAddresses = sort validatorsAddresses where address1 < adress2 if (roundHash(address1) < roundHash(address2))
+                                 or ((roundHash(address1) == roundHash(address2)) and address1 < address2)         
+    
+    return shuffledValidatorAddresses
+```
+
+
+### Block Processing
+
+
+#### After Genesis Block Execution
+
+After the genesis block `b` is executed, the following logic is executed:
+
+```python
+create the entry in the genesis data substore with 
+store value = {
+    "height": b.header.height, 
+    "initRounds": b.header.assets.initRounds, 
+    "initDelegates": b.header.assets.initDelegates
+} serialized using genesisDataStoreSchema
+
+# set the inititial delegates in the BFT module
+bftWeights = [
+    {"address": address, "bftWeight": 1} 
+    for address in initDelegates
+    sorted by lexicographically by address
+]
+
+# compute the initial BFT threshold 
+initBFTThreshold = floor(2/3 * length(initDelegates)) + 1
+
+BFT.setBFTParameters(initBFTThreshold,
+                     initBFTThreshold,
+                     bftWeights)
+                     
+# set the inititial delegates in the validators module                 
+validators.setGeneratorList(initDelegates)
+```
+
+
+#### After Block Execution
+
+After a block `newBlock` is executed, the properties related to missed blocks are updated according to [Delegate Productivity][LIP-0023-delegateProductivity]. This logic is recapitulated below:
+
+```python
+newHeight = newBlock.header.height
+# previousTimestamp is the value in the previous timestamp substore
+missedBlocks = validators.getGeneratorsBetweenTimestamps(previousTimestamp, newBlock.header.timestamp)
+
+# remove the start and end blocks, as those are not missed
+missedBlocks[validators.getGeneratorAtTimestamp(previousTimestamp)] -= 1
+missedBlocks[validators.getGeneratorAtTimestamp(newBlock.header.timestamp)] -= 1
+
+for address in missedBlocks:
+    delegateStore(address).consecutiveMissedBlocks += missedBlocks[address]
+
+    # the rule below was introduced in LIP 0023
+    if (delegateStore(address).consecutiveMissedBlocks > FAIL_SAFE_MISSED_BLOCKS
+        and newHeight - delegateStore(address).lastGeneratedHeight > FAIL_SAFE_INACTIVE_WINDOW):
+        delegateStore(address).isBanned = True
```

Comment:
question: basically u can still get banned after generating the block (if in the meanwhile you missed a lot of blocks). Is the block still valid in this case?

Reply:
yes